plugin/plugins/filesystem/nodes/utils.go

```diff
@@ -0,0 +1,197 @@
+		attr.Mode |= unixfsTypeTo9Mode(ufsNode.Type())
+	}
+
+	if attrMask.Blocks {
+		if bs := ufsNode.BlockSizes(); len(bs) != 0 {
+			attr.BlockSize = bs[0] //NOTE: this value is to be used as a hint only; subsequent child block size may differ
+		}
+
+		//TODO [eventualy]: switch off here for handling of time metadata in new format standard
+		timeStamp(attr, attrMask)
+	}
+
+	if attrMask.Size {
+		attr.Size, attrMask.Size = ufsNode.FileSize(), true
+	}
+
+	if attrMask.RDev {
+		switch path.Namespace() {
+		case "ipfs":
+			attr.RDev, attrMask.RDev = dIPFS, true
+			//case "ipns":
+			//attr.RDev, attrMask.RDev = dIPNS, true
+			//etc.
+		}
+	}
+
+	return nil
+}
+
+func ipldStat(dirEnt *p9.Dirent, node ipld.Node) error {
+	ufsNode, err := unixfs.ExtractFSNode(node)
+
+	if err != nil {
+		return err
+	}
+
+	nodeType := unixfsTypeTo9Mode(ufsNode.Type()).QIDType()
+
+	dirEnt.Type = nodeType
+	dirEnt.QID.Type = nodeType
+	dirEnt.QID.Path = cidToQPath(node.Cid())
+
+	return nil
+}
+
+func cidToQPath(cid cid.Cid) uint64 {
+	hasher := fnv.New64a()
+	if _, err := hasher.Write(cid.Bytes()); err != nil {
+		panic(err)
+	}
+	return hasher.Sum64()
+}
+
+func coreTypeTo9Mode(ct coreiface.FileType) p9.FileMode {
+	switch ct {
+	// case coreiface.TDirectory, unixfs.THAMTShard // Should we account for this?
+	case coreiface.TDirectory:
+		return p9.ModeDirectory
+	case coreiface.TSymlink:
+		return p9.ModeSymlink
+	default: //TODO: probably a bad assumption to make
+		return p9.ModeRegular
+	}
+}
+
+//TODO: see if we can remove the need for this; rely only on the core if we can
+func unixfsTypeTo9Mode(ut unixpb.Data_DataType) p9.FileMode {
+	switch ut {
+	// case unixpb.Data_DataDirectory, unixpb.Data_DataHAMTShard // Should we account for this?
+	case unixpb.Data_Directory:
+		return p9.ModeDirectory
+	case unixpb.Data_Symlink:
+		return p9.ModeSymlink
+	default: //TODO: probably a bad assumption to make
+		return p9.ModeRegular
+	}
+}
+
+func coreEntTo9Ent(coreEnt coreiface.DirEntry) p9.Dirent {
+	entType := coreTypeTo9Mode(coreEnt.Type).QIDType()
+
+	return p9.Dirent{
+		Name: coreEnt.Name,
+		Type: entType,
+		QID: p9.QID{
+			Type: entType,
+			Path: cidToQPath(coreEnt.Cid)}}
+}
+
+const ( // pedantic POSIX stuff
+	S_IROTH p9.FileMode = p9.Read
+	S_IWOTH             = p9.Write
+	S_IXOTH             = p9.Exec
+
+	S_IRGRP = S_IROTH << 3
+	S_IWGRP = S_IWOTH << 3
+	S_IXGRP = S_IXOTH << 3
+
+	S_IRUSR = S_IRGRP << 3
+	S_IWUSR = S_IWGRP << 3
+	S_IXUSR = S_IXGRP << 3
+
+	S_IRWXO = S_IROTH | S_IWOTH | S_IXOTH
+	S_IRWXG = S_IRGRP | S_IWGRP | S_IXGRP
+	S_IRWXU = S_IRUSR | S_IWUSR | S_IXUSR
+
+	IRWXA = S_IRWXU | S_IRWXG | S_IRWXO            // 0777
+	IRXA  = IRWXA &^ (S_IWUSR | S_IWGRP | S_IWOTH) // 0555
+)
+
+func defaultRootAttr() (attr p9.Attr, attrMask p9.AttrMask) {
+	attr.Mode = p9.ModeDirectory | IRXA
+	attr.RDev = dMemory
+	attrMask.Mode = true
+	attrMask.RDev = true
+	attrMask.Size = true
+	//timeStamp(&attr, attrMask)
```

Comment:
Yeah, this should probably have a time parameter. We can store the current time on filesystem instantiation, and pass it in rather than using the current time or 0.